File: pdv-julia/src/modules.jl
# modules.jl — Handler registry and dispatch for custom type handlers.
#
# Port of pdv/modules.py. On a Julia kernel there are two registration paths:
#
# 1. **Multiple dispatch** (the idiomatic path, used by the bundled
#    N-pendulum-julia module): a module adds a method to PDVKernel's
#    `pdv_handle` generic function:
#
#        import PDVKernel: pdv_handle
#        function pdv_handle(obj::MyType, path::String, tree::PDVTree) ... end
#
# 2. **Explicit registration** (`register_handler(f, T)`) for wrapping types
#    the caller does not own. Registered handlers win over `pdv_handle`
#    methods, mirroring Python's registered-over-dunder precedence.
#
# Built-in default handlers (Python's default_handlers.py) live in
# default_handlers.jl and register lazily from the lookup functions below —
# numeric arrays plot via whatever Makie backend is loaded, Makie figures
# and DataFrames display, and user registrations always win over defaults.

"""One explicitly registered handler mapping a type to a callable."""
struct HandlerEntry
    type::Type
    func::Function
    type_name::String
end

const _HANDLER_REGISTRY = Dict{Type,HandlerEntry}()

"""
    register_handler(func, T::Type)

Register `func(obj, path, tree)` as the double-click handler for values of
type `T` and its subtypes — concrete registrations match via the supertype
walk, abstract/parametric ones via an `isa` scan (most-specific-wins). The
explicit-registration analog of adding a `pdv_handle` method; registered
handlers take precedence.
"""
function register_handler(func::Function, T::Type)
    type_name = fully_qualified_type_name(T)
    if haskey(_HANDLER_REGISTRY, T)
        @warn "Handler for $type_name overwritten"
    end
    _HANDLER_REGISTRY[T] = HandlerEntry(T, func, type_name)
    nothing
end

# Walk the type hierarchy for a registered entry. The exact-key supertype
# walk covers concrete registrations cheaply, but abstract or parametric
# (UnionAll) registrations — `register_handler(f, AbstractVector)`,
# `register_handler(f, Spectrum)` where `Spectrum{T}` is parametric — never
# appear in a concrete type's supertype chain (`Vector{Float64}` walks
# through `AbstractVector{Float64}`, not `AbstractVector`), so the docstring
# promise "and subtypes" silently never matched them (second review;
# pdv-python's `__mro__` walk delivers base-class matches). Those fall back
# to an `isa` scan, most-specific-wins.
function _registered_handler_for(obj)
    T = typeof(obj)
    while true
        entry = get(_HANDLER_REGISTRY, T, nothing)
        entry !== nothing && return entry
        T === Any && break
        T = supertype(T)
    end
    best = nothing
    for entry in values(_HANDLER_REGISTRY)
        obj isa entry.type || continue
        if best === nothing || entry.type <: best.type
            best = entry
        end
    end
    return best
end

"""
    has_handler_for(obj) -> Bool

True when an explicitly registered handler matches `obj`'s type, or a
`pdv_handle` method exists for it.
"""
function has_handler_for(obj)::Bool
    register_default_handlers!()
    _registered_handler_for(obj) !== nothing && return true
    return hasmethod(pdv_handle, Tuple{typeof(obj),String,PDVTree})
end

"""
    dispatch_handler(obj, path, tree) -> Dict

Find and call the handler for `obj`. Exceptions are caught and returned as
`{"dispatched" => false, "error" => "..."}` so the renderer receives a
structured error rather than an opaque kernel exception.
"""
function dispatch_handler(obj, path::AbstractString, tree)::Dict{String,Any}
    register_default_handlers!()
    entry = _registered_handler_for(obj)
    if entry !== nothing
        try
            Base.invokelatest(entry.func, obj, String(path), tree)
        catch err
            return Dict{String,Any}(
                "dispatched" => false,
                "error" => "Handler for $(entry.type_name) failed: $(sprint(showerror, err))")
        end
        return Dict{String,Any}("dispatched" => true)
    end
    if hasmethod(pdv_handle, Tuple{typeof(obj),String,PDVTree})
        try
            Base.invokelatest(pdv_handle, obj, String(path), tree)
        catch err
            return Dict{String,Any}(
                "dispatched" => false,
                "error" => "pdv_handle failed: $(sprint(showerror, err))")
        end
        return Dict{String,Any}("dispatched" => true)
    end
    return Dict{String,Any}(
        "dispatched" => false,
        "error" => "No handler for $(julia_type_string(obj))")
end

"""
    get_handler_registry() -> Dict{String,String}

Return `{type_name => handler_name}` for explicitly registered handlers plus
every user-defined `pdv_handle` method.
"""
function get_handler_registry()::Dict{String,String}
    registry = Dict{String,String}(
        entry.type_name => string(entry.func) for entry in values(_HANDLER_REGISTRY))
    for m in methods(pdv_handle)
        sig = Base.unwrap_unionall(m.sig)
        sig isa DataType && length(sig.parameters) >= 2 || continue
        T = sig.parameters[2]
        T isa DataType || continue
        registry[fully_qualified_type_name(T)] = "pdv_handle"
    end
    return registry
end

"""Clear the explicit handler registry and the lazy-defaults latch (tests)."""
clear_handlers!() = (empty!(_HANDLER_REGISTRY); _reset_default_handlers!(); nothing)

File: pdv-python/pdv/modules.py
"""
pdv.modules — Handler registry, decorator, and dispatch for custom type handlers.

Module developers use ``@pdv.handle(MyClass)`` to register a handler function
that is invoked when a user double-clicks a tree node whose value is an instance
of ``MyClass``. Handlers are resolved by walking the MRO, so a handler
registered on a base class also applies to subclasses.

Classes the user authors directly may instead define a ``__pdv_handle__``
method (the dunder protocol — see :mod:`pdv.serializers`); the registered
handler wins when both are present. ``dispatch_handler`` wraps both paths in
try/except so a handler exception surfaces to the renderer as a structured
``{"dispatched": False, "error": "..."}`` reply rather than an opaque kernel
error.

Public API
----------
handle : decorator factory
    ``@pdv.handle(MyClass)`` registers a handler for instances of MyClass.
has_handler_for : function
    Check whether any registered handler matches an object's type (via MRO),
    or the value's class defines ``__pdv_handle__``.
dispatch_handler : function
    Find and call the appropriate handler for an object. Catches handler
    exceptions and returns them as a structured error.
get_handler_registry : function
    Return a snapshot of all registered handlers.
clear_handlers : function
    Clear the handler registry (used in tests).

See Also
--------
ARCHITECTURE.md §3.4 (message type catalogue)
pdv.serializers (dunder protocol surface, including ``__pdv_handle__``)
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable


@dataclass
class HandlerEntry:
    """One registered handler mapping a type to a callable."""

    cls: type
    func: Callable
    class_name: str  # fully qualified: "module.Class"


_handler_registry: dict[type, HandlerEntry] = {}


def handle(cls: type) -> Callable:
    """Decorator factory: register a handler for instances of *cls*.

    Usage::

        @pdv.handle(MyClass)
        def on_my_class(obj, path, pdv_tree):
            ...

    Parameters
    ----------
    cls : type
        The type to register a handler for.

    Returns
    -------
    Callable
        A decorator that registers the wrapped function.
    """

    def decorator(func: Callable) -> Callable:
        class_name = f"{cls.__module__}.{cls.__qualname__}"
        if cls in _handler_registry:
            import warnings

            old = _handler_registry[cls]
            warnings.warn(
                f"Handler for {class_name} overwritten (was {old.func.__qualname__})"
            )
        _handler_registry[cls] = HandlerEntry(cls=cls, func=func, class_name=class_name)
        return func

    return decorator


def has_handler_for(obj: Any) -> bool:
    """Check whether any handler matches *obj* — registered or dunder.

    Looks up ``@pdv.handle``-registered handlers via MRO walk, then falls
    back to checking whether ``type(obj)`` defines ``__pdv_handle__``.

    Parameters
    ----------
    obj : Any
        The object to check.

    Returns
    -------
    bool
        True if a handler is registered for ``type(obj)`` or any of its bases,
        or if the class defines a ``__pdv_handle__`` method.
    """
    for cls in type(obj).__mro__:
        if cls in _handler_registry:
            return True
    return hasattr(type(obj), "__pdv_handle__")


def dispatch_handler(obj: Any, path: str, pdv_tree: Any) -> dict:
    """Find and call the handler for *obj*.

    Resolution order: registered ``@pdv.handle`` (MRO walk) wins over a
    class-defined ``__pdv_handle__`` dunder. Exceptions from either path
    are caught and returned as ``{"dispatched": False, "error": "..."}`` so
    the renderer receives a structured error rather than an opaque kernel
    exception.

    Parameters
    ----------
    obj : Any
        The tree node value.
    path : str
        Dot-separated tree path of the node.
    pdv_tree : Any
        The PDVTree instance.

    Returns
    -------
    dict
        ``{"dispatched": True}`` on success, or
        ``{"dispatched": False, "error": "..."}`` when no handler matches
        or the chosen handler raised.
    """
    for cls in type(obj).__mro__:
        if cls in _handler_registry:
            try:
                _handler_registry[cls].func(obj, path, pdv_tree)
            except Exception as exc:  # noqa: BLE001
                return {
                    "dispatched": False,
                    "error": (
                        f"Handler for {cls.__module__}.{cls.__qualname__} "
                        f"failed: {exc}"
                    ),
                }
            return {"dispatched": True}
    if hasattr(type(obj), "__pdv_handle__"):
        try:
            obj.__pdv_handle__(path, pdv_tree)
        except Exception as exc:  # noqa: BLE001
            return {"dispatched": False, "error": f"__pdv_handle__ failed: {exc}"}
        return {"dispatched": True}
    t = type(obj)
    return {
        "dispatched": False,
        "error": f"No handler for {t.__module__}.{t.__qualname__}",
    }


def get_handler_registry() -> dict[str, str]:
    """Return ``{class_name: handler_func_name}`` for all registered handlers.

    Returns
    -------
    dict[str, str]
        Mapping from fully qualified class name to handler function qualname.
    """
    return {
        entry.class_name: entry.func.__qualname__
        for entry in _handler_registry.values()
    }


def clear_handlers() -> None:
    """Clear the handler registry. Used in tests."""
    _handler_registry.clear()
